contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,508 @@
+                    minter: from_vec(env, decoded.minter.as_ref()),
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+#[allow(dead_code)]
+impl HubMessage {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::SendToHub {
+                destination_chain,
+                message,
+            } => SendToHub {
+                messageType: MessageType::SendToHub.into(),
+                destination_chain: to_std_string(destination_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+            Self::ReceiveFromHub {
+                source_chain,
+                message,
+            } => ReceiveFromHub {
+                messageType: MessageType::ReceiveFromHub.into(),
+                source_chain: to_std_string(source_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let payload = payload.to_alloc_vec();
+
+        let message_type = MessageType::abi_decode(&payload[0..32], true)
+            .map_err(|_| MessageError::InvalidMessageType)?;
+
+        let message = match message_type {
+            MessageType::SendToHub => {
+                let decoded = SendToHub::abi_decode_params(&payload, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::SendToHub {
+                    destination_chain: String::from_str(env, &decoded.destination_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                }
+            }
+            MessageType::ReceiveFromHub => {
+                let decoded = ReceiveFromHub::abi_decode_params(&payload, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::ReceiveFromHub {
+                    source_chain: String::from_str(env, &decoded.source_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                }
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+fn to_std_string(soroban_string: String) -> StdString {
+    let length = soroban_string.len() as usize;
+    let mut bytes = vec![0u8; length];
+    soroban_string.copy_into_slice(&mut bytes);
+    StdString::from_utf8(bytes).expect("Invalid UTF-8 sequence")
```

Comment:
Since the `copy_into_slice` method panics if the output slice and string are of different lengths, is that guarantee enough that the conversion is lossless since it's basically a byte-level copy?